Track meters during playback

embarc-gabriel opened the pull request:
Resolves: #8705 

Implement track meters during playback. See issue for Figma link.

The most challenging aspect of this PR involves modifying the AudioIO files in the lib-audio-io module to enable retrieving meter information on a per-track basis.
Previously, all inputs were mixed on the producer side, with only the final mixed output being added to the ring buffer.
Now, each track has its own dedicated ring buffer.

While we could potentially move the mixing process to the consumer side and utilize these per-track ring buffers, I'm hesitant to make extensive changes to this file at this time to minimize the risk of introducing regressions. A future refactoring ticket could address this.

Otherwise, the solution mirrors the approach taken for the main meter. We've introduced a TrackMeter and integrated it into the audiooutput component. Other code sections can subscribe to a channel to monitor changes in the track meter data.

<!-- Use "x" to fill the checkboxes below like [x] -->

- [x] I signed [CLA](https://www.audacityteam.org/cla/)
- [x] The title of the pull request describes an issue it addresses
- [x] If changes are extensive, then there is a sequence of easily reviewable commits
- [x] Each commit's message describes its purpose and effects
- [x] There are no behavior changes unnecessary for the stated purpose of the PR

Recommended:
- [ ] Each commit compiles and runs on my machine without known undesirable changes of behavior

To QA:
- [x] Stop clears max peak and recent peak information but keeps the clip marker if the signal was clipped before.
- [x] Play clears peaks and also previous clipped information
- [x] Click on the track meter clears peaks and previous clipped information
- [x] Recent peak information consider the last 600ms of data
- [x] During clipping the bar goes full red
- [x] Works with realtime effects that have algorithmic delay (e.g. https://www.thimeo.com/stereo-tool/)
- [x] Stereo tracks with noticeably different left/right levels are represented correctly by the meters
- [x] Works when adding/removing realtime effects during playback

embarc-gabriel commented on au3/libraries/lib-audio-devices/IMeterChannel.h:
```diff
@@ -0,0 +1,15 @@
+*/
+#pragma once
+
+constexpr int64_t MAIN_METER_KEY = -1;
```
My decision here was to prioritize simplicity. While we could have introduced greater complexity to distinguish the toolbar meter from the track meters in the type system, I find this simpler approach more intuitive.

saintmatthieu replied:
I used a similar mechanism for the realtime effects, to uniquely identify the master track, only I used `-2` because `-1` is often used to denote uninitialized value. It's worked well.
The naming has confused me, though. I think the concept of "master track" applies here, too. Hence I'd suggest `MASTER_TRACK_ID` instead.

embarc-gabriel replied:
Done

embarc-gabriel commented on au3/libraries/lib-audio-devices/IMeterChannel.h:
```diff
@@ -0,0 +1,15 @@
+    virtual void push(uint8_t channel, float signal, int64_t key = MAIN_METER_KEY) = 0;
+    virtual void sendAll() = 0;
+    virtual void reset() = 0;
+};
```
I haven't included this in the IoC because this interface is injected into the Au3 code. My intention is to keep the Au4 code separate from the Au3 code. For this same reason, the channel reading functionality isn't part of the abstract class, as I've decided to use channels for the current solution. In some ways, the term "channel" might be misleading, as we can only send data through it using this interface. I could potentially work around the IoC using multiple inheritance and a pointer cast, but again, I've opted for simplicity. I'm open to suggestions on this.

embarc-gabriel commented on au3/libraries/lib-audio-io/AudioIO.cpp:
```diff
@@ -2943,31 +3024,66 @@ void AudioIoCallback::SendVuOutputMeterData(
-    //if (pMixerBoard)
-    //   pMixerBoard->UpdateMeters(GetStreamTime(),
-    //                              (pProj->GetControlToolBar()->GetLastPlayMode() == loopedPlay));
+    PushMainMeterValues(outputMeter, outputMeterFloats, mNumPlaybackChannels, framesPerBuffer);
```
Here is where "devil" allocations and mutex lock takes place.
This must be improved in the future as discussed before.

embarc-gabriel replied:
Edit: Probably I can easily allocate the memory outside the audio thread. I can work on this on the next ticket.

saintmatthieu replied:
Thanks for making those comments about critical aspects of your PR, I'd have missed some important things otherwise.

embarc-gabriel replied:
Yeah! @IliasBergstromMuse inspired me to do this more often.

saintmatthieu commented on au3/libraries/lib-audio-io/AudioIO.cpp:
```diff
@@ -2943,31 +3024,66 @@ void AudioIoCallback::SendVuOutputMeterData(
+                frames
+            );
+    
+            float peak = 0.0f;
```
... and actually, if you did that, you'd also be led to removing the use of `std::vector<float> peak(channels, -1.0f);` above, which is one memory allocation.

saintmatthieu commented on src/playback/internal/au3/au3audiooutmeter.cpp:
```diff
@@ -0,0 +1,56 @@
+#include "libraries/lib-utility/MemoryX.h"
+
+namespace au::playback {
+OutMeter::OutMeter()
```
We could probably easily reduce the risk of memory allocation by reserving a generous amount of slots in `m_trackData`.
I guess that Au3 is using lock-free queues to transmit volume values from the audio driver thread to the main thread, and that it pre-allocates the number of slots for this. Looking at how many it pre-allocates should be re-usable here.

embarc-gabriel replied:
Done.

saintmatthieu reviewed (commented):
Finished with this first iteration. Comments I'd like to see responded to or addressed:
* `mNumPlaybackChannels`
* `MASTER_TRACK_ID`
* remove vector in `PushMainMeterValues`
* reserving memory in `OutMeter::OutMeter()`

Also: you add this new `mPlaybackTracks` member to `AudioIO`. Have you considered reusing the `mProcessingBuffers` to get the per-track volume values from?

embarc-gabriel commented on au3/libraries/lib-audio-io/AudioIO.h:
```diff
@@ -203,6 +215,8 @@ class AUDIO_IO_API AudioIoCallback /* not final */ : public AudioIOBase
         const float* inputSamples, unsigned long framesPerBuffer);
     void SendVuOutputMeterData(
         const float* outputMeterFloats, unsigned long framesPerBuffer);
+    void PushMainMeterValues(std::shared_ptr<IMeterChannel> channel, const float* values, uint8_t channels, unsigned long frames);
```
It cannot be const once I am adding data to the channel.
But yes, it could be reference.
Usually I prefer to pass it by copy to enforce the shared ownership aspect but I have read more on this today and maybe a reference fits best here. Thanks.

saintmatthieu replied:
I think `const std::shared_ptr<IMeterChannel>&` would work (what would not is `std::shared_ptr<const IMeterChannel>`) - please give it a try.

embarc-gabriel commented on src/projectscene/view/trackspanel/trackitem.cpp:
```diff
@@ -34,13 +37,31 @@ TrackItem::TrackItem(QObject* parent)
 TrackItem::~TrackItem()
 {
-    // m_audioSignalChanges.resetOnReceive(this);
+    m_playbackTrackSignalChanged.close();
```
IMO once we remove the track the channel for that track should be close. The meter receives this information and remove the channel from the list of channels. This seems to me a nice cleanup.
In the previous implementation we don´t have a channel per track.

embarc-gabriel replied:
I will close it but you can reopen and we can discuss it further.

embarc-gabriel commented on au3/libraries/lib-audio-io/AudioIO.cpp:
```diff
@@ -2171,25 +2211,48 @@ bool AudioIO::ProcessPlaybackSlices(
+                        samplesAvailable, 0);
                 }
             }
+            
```
Fixed.

embarc-gabriel commented on src/projectscene/qml/Audacity/ProjectScene/trackspanel/audio/VolumePressureMeter.qml:
```diff
@@ -202,29 +263,52 @@ Canvas {
-        const overloadStyle = root.isClipping ? "#EF476F" : ui.theme.buttonColor
+        const overloadStyle = prv.clipped ? "#EF476F" : ui.theme.buttonColor
         drawRoundedRect(ctx, overloadStyle, 0, 0, indicatorWidth, prv.overloadHeight, 2, "top")
+        
```
Fixed.

embarc-gabriel commented on src/projectscene/view/trackspanel/trackitem.cpp:
```diff
@@ -34,13 +37,31 @@ TrackItem::TrackItem(QObject* parent)
 
+    m_playbackTrackSignalChanged = playback()->audioOutput()->playbackTrackSignalChanges(m_trackId);
+    m_playbackTrackSignalChanged.onReceive(this, [this](au::audio::audioch_t channel, const au::audio::AudioSignalVal& newValue) {
+        const float newPressure = std::min(std::max(newValue.pressure, MIN_ALLOWED_PRESSURE), MAX_ALLOWED_PRESSURE);
```
True! Done!

embarc-gabriel commented on src/record/CMakeLists.txt:
```diff
@@ -40,7 +40,7 @@ include(${CMAKE_CURRENT_LIST_DIR}/../au3wrap/au3defs.cmake)
 set(MODULE_INCLUDE ${AU3_INCLUDE})
 set(MODULE_DEF ${AU3_DEF})
-set(MODULE_LINK au3wrap)
+set(MODULE_LINK au3wrap playback)
```
au3audioinoutmeter is on playback module and is used on the record module.
The previous code was forwarding declaring this class that's why it was compiling, I think.
The next PR will probably change this.
My idea is to add the meter on the au3wrap and use it on both playback and record modules.

embarc-gabriel commented on au3/libraries/lib-audio-io/AudioIO.cpp:
```diff
@@ -1346,6 +1366,19 @@ bool AudioIO::AllocateBuffers(
                 mPlaybackSchedule.mTimeQueue.Init(timeQueueSize);
             }
 
+            const size_t mNumPlaybackChannels = GetNumPlaybackChannels();
```
Yes! Sure! Fixed.

embarc-gabriel commented:
> Finished with this first iteration. Comments I'd like to see responded to or addressed:
> 
>     * `mNumPlaybackChannels`
> 
>     * `MASTER_TRACK_ID`
> 
>     * remove vector in `PushMainMeterValues`
> 
>     * reserving memory in `OutMeter::OutMeter()`
> 
> 

@saintmatthieu I have addressed all comments. If there is anything else let me know.

embarc-gabriel commented on au3/libraries/lib-audio-io/AudioIO.cpp:
```diff
@@ -1291,6 +1307,11 @@ bool AudioIO::AllocateBuffers(
+
+                    auto outputMeter = mOutputMeter.lock();
+                    if (outputMeter) {
+                        outputMeter->reserve(mPlaybackSequences.size() * mNumPlaybackChannels);
```
yes! Good catch!

embarc-gabriel replied:
Done!

igorkorsukov commented on src/playback/internal/au3/au3audiooutmeter.cpp:
```diff
@@ -0,0 +1,61 @@
+void OutMeter::sendAll()
+{
+    m_audioSignalChanges.send(std::move(m_trackData));
+    m_trackData.clear();
```
I think it's not possible to make a class member's move and then continue to address it...

embarc-gabriel replied:
You are right! Thank you! Fixed!
